### PR TITLE
VZ-5710 update istio ingress type/ports

### PR DESCRIPTION
**Pull request by guoyongzhang**
# Description

- add service type and ports to spec.istio.ingress in Verrazzano CRD
- support NodePort and ports override in istio ingress by adding type and ports to istio operator yaml 
- externalURLs validation with NodePort is separated for nginx and istio now that they have their own service type
- In addition to validateUpdate(), the same validation logic is also added to validateInstall() in nginx and istio
- In application URLs, ExternalIPs takes precedence over LoadBalancer IP
- When istio ingress is set with NodePort, Application URLs use ExternalIPs instead of pod host ip
- replace the hard-coded ip in tests with randomly generated fake ip so sonar doesn't complain

Fixes VZ-5709, VZ-5710, VZ-5711, VZ-5712

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 1]
```diff
@@ -136,14 +135,37 @@ func (i istioComponent) Name() string {
...
+	if vz.Components.Istio.Ingress.Type != vzapi.NodePort {
+		return nil
+	}
+
+	// look for externalIPs if NodePort
+	if vz.Components.Istio.Ingress.Type == vzapi.NodePort {
+		if vz.Components.Istio.IstioInstallArgs != nil {
+			return vzconfig.CheckArgs(vz.Components.Istio.IstioInstallArgs, ExternalIPArg, i.Name())
+		}
+		return fmt.Errorf("'istioInstallArgs' cannot be empty. ExternalIPs needs to be specified here as type is NodePort")
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/istio/istio_component.go:166]: I think `ExternalIPArg` needs to be named in the error message, something like

```
return fmt.Errorf("%s must be specified in 'istioInstallArgs' for NodePort ingress type", ExternalIPArg)
```

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:166]: can be done as a follow-up

File: platform-operator/controllers/verrazzano/component/nginx/nginx_component.go [hunk 2]
```diff
@@ -78,10 +80,34 @@ func (c nginxComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Verrazz
...
+		return nil
+	}
+
+	// look for externalIPs if NodePort
+	if vz.Components.Ingress.Type == vzapi.NodePort {
+		if vz.Components.Ingress.NGINXInstallArgs != nil {
+			return vzconfig.CheckArgs(vz.Components.Ingress.NGINXInstallArgs, nginxExternalIPKey, c.Name())
+		}
+		// if ingress args are not set at all, then external ips will not be set
+		return fmt.Errorf("'nginxInstallArgs' cannot be empty. ExternalIPs needs to be specified here as type is NodePort")
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/nginx/nginx_component.go:109]: Same comment re: error message.

File: platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go [hunk 3]
```diff
@@ -543,11 +543,19 @@ type IngressNginxComponent struct {
 
 // IstioIngressSection specifies the specific config options available for the Istio Ingress Gateways.
 type IstioIngressSection struct {
+	// Type of ingress.  Default is LoadBalancer
+	// +optional
+	Type IngressType `json:"type,omitempty"`
+	// Ports to be used for NGINX
```
Comment (michael-cico) [inline platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go:549]: Typo, NGINX --> Istio Ingress Gateway

File: platform-operator/internal/vzconfig/args.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	"fmt"
+	"net"
+
+	vzapi "github.com/verrazzano/verrazzano/platform-operator/apis/verrazzano/v1alpha1"
+)
+
+// This method goes through the key-value pair to detect the presence of the
+// specific key for the corresponding component that holds the external IP address
+// It also checks whether IP addresses are valid and provided in a List format
+func CheckArgs(installArgs []vzapi.InstallArgs, argsKeyName, compName string) error {
```
Comment (michael-cico) [inline platform-operator/internal/vzconfig/args.go:16]: Function name is general, however this looks specific to checking for ExternalIP args.  Should probably reflect that in the name. (follow-up PR is OK if you want to change it later)

**Review by michael-cico — APPROVED** [review]

File: platform-operator/internal/vzconfig/args.go (hunk 4 above)
Comment (guoyongzhang) [reply platform-operator/internal/vzconfig/args.go:16]: the function is general for looking for specified argsKeyName.

File: platform-operator/internal/vzconfig/args.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	var keyPresent bool
+	for _, installArg := range installArgs {
```
[the remark this answers is not in the document]
Comment (guoyongzhang) [reply platform-operator/internal/vzconfig/args.go]: fixed

File: platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go (hunk 3 above)
Comment (guoyongzhang) [reply platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go]: fixed

File: platform-operator/controllers/verrazzano/component/nginx/nginx_component.go (hunk 2 above)
Comment (guoyongzhang) [reply platform-operator/controllers/verrazzano/component/nginx/nginx_component.go]: fixed

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go (hunk 1 above)
Comment (guoyongzhang) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go]: fixed

**Review by michael-cico — APPROVED** [review]

File: platform-operator/internal/vzconfig/args.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2021, 2022, Oracle and/or its affiliates.
```
Comment (kminder) [inline platform-operator/internal/vzconfig/args.go:1]: Is this new or modified?

File: platform-operator/internal/vzconfig/args.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+// This method goes through the key-value pair to detect the presence of the
+// specific key for the corresponding component that holds the external IP address
+// It also checks whether IP addresses are valid and provided in a List format
+func CheckArgs(installArgs []vzapi.InstallArgs, argsKeyName, compName string) error {
+	var keyPresent bool
+	for _, installArg := range installArgs {
+		if installArg.Name == argsKeyName {
+			keyPresent = true
+			if len(installArg.ValueList) < 1 {
+				return fmt.Errorf("At least one %s external ips need to be set as an array for the key \"%v\"", compName, installArg.Name)
```
Comment (kminder) [inline platform-operator/internal/vzconfig/args.go:22]: Should 'ips' be 'IPs'

File: platform-operator/internal/vzconfig/args.go (hunk 6 above)
Comment (guoyongzhang) [reply platform-operator/internal/vzconfig/args.go]: fixe

File: platform-operator/internal/vzconfig/args.go (hunk 7 above)
Comment (guoyongzhang) [reply platform-operator/internal/vzconfig/args.go]: fixed

**Review by michael-cico — APPROVED** [review]

**Pull request merged**